0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,395 @@
+
+For example, one cent represents an asset scale of 2 in the case of USD, whereas one satoshi represents an asset scale of 8 in the case of Bitcoin.
+
+### Selecting scales
+
+Settlement engines are RECOMMENDED to perform or fulfill requests with amounts denominated in the same scale as its settlements.
+
+Account balances within the accounting systems are RECOMMENDED to be denominated in the same scale as their corresponding settlement engine, but MAY use different ones. For example, micropayments may require more precision than can actually be settled, or databases may limit precision to less than the settlement system is capable of.
+
+### **`Quantity`** object
+
+An amount denominated in some unit of a single, fungible asset. (Since each account is denominated in a single asset, the type of asset is implied.)
+
+#### Attributes
+
+- **`amount`** &mdash; string
+  - Amount of the unit, which is a non-negative integer.
+  - This amount is encoded as a string to ensure no precision is lost on platforms that don't natively support arbitrary precision integers.
+- **`scale`** &mdash; number
+  - Asset scale of the unit, between `0` and the maximum 8-bit unsigned integer, `255` (inclusive).
+
+#### Example
+
+To represent $2.54 in units of cents, where the amount is a multiple of $0.01:
+
+```json
+{
+  "amount": "254",
+  "scale": 2
+}
+```
+
+### Scale conversions
+
+If the accounting system or settlement engine receives a request with a **[`Quantity`](#quantity-json-type)** denominated in a unit more precise than its unit, it MUST convert the quantity into its native unit. If so, the resulting amount MUST be rounded down before fulfilling the request.
+
+The response to the request MUST include the converted, rounded **[`Quantity`](#quantity-json-type)** used to fulfill the request, which MUST be less than or equal to the amount sent in the original request. (If the amount rounds down to 0, then the amount in the response would be 0.)
+
+Then, the system with the additional precision initiating the request MUST track the leftover sum so it may accumulate and be retried in subsequent requests.
+
+## Settlement Engine HTTP API
+
+_Connector &rarr; settlement engine_
+
+HTTP/2 is RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Setup an account
+
+Informs the settlement engine that a new account was created within the accounting system using the given [account identifier](#accounts-and-identifiers). The settlement engine MAY perform tasks as a prerequisite to settle with the account. For example, a settlement engine implementation might send messages to the peer to exchange ledger identifiers or to negotiate settlement-related fees.
+
+#### Request
+
+```http
+PUT /accounts/:id HTTP/1.1
```

Comment:
@adrianhopebailie @sappenin had a discussion on this here: https://interledger.slack.com/archives/CHZC9MB88/p1565629683255200?thread_ts=1565022687.213900&cid=CHZC9MB88

> This PR for the XRPL Settlement engine illustrates this "middle ground": if the client supplies and Id, the SE will use it. If the client doesn't supply an Id, the SE will generate one using Type4 UUID): https://github.com/interledgerjs/settlement-xrp/pull/10 (edited)

Reply:
IMHO, I don't think it's too difficult for SE's to support both patterns (i.e., "create with POST" _and_ "create/update with PUT").

To that end, I vote we have two endpoints:

* POST `/accounts` (create a new account, SE generates the accountId).
* PUT `/accounts/:id`(create an account with an id; or update an existing account by its id)

I'm pretty sure we need a `PUT` endpoint anyway in order to update callback URLs on an account, and adding a 2nd endpoint for `POST` is trivial. 

This way we don't bikeshed on this any further, _and_ we support varying styles of API design so that clients and servers can decide how best to manage their identifier space.

Reply:
Let's just stick with PUT only. Make POST optional.
PUT as a create is legal and the smaller we keep the API surface the better IMO

Reply:
Agree with @adrianhopebailie  on all points. As long as the spec doesn't preclude my proposal (which, in it's current form it does not), then I'm supportive of including only `PUT` here in the RFC.

I'm going to resolve this conversation because I don't think there are any changes required here.